### DISCUSSION TITLE
Gradient with respect to RV

### DISCUSSION THREAD

**Discussion by pecey** [jax-ml/jax#14581] [Q&A]
It is possible to write sampling functions in two ways. In one case, the key is passed into the function and the sampling happens inside the function. And in another case, the sampling happens outside the function.
```
def fn_key(key):
    return 3 + jax.random.normal(key) * 2

eps = jax.random.normal(key)
def fn_eps(eps):
    return 3 + eps * 2
```

Computing the grad is quite obvious in the second case. I was wondering how to write the same for the first case (or even if it is possible).
I was trying to write a `custom_jvp`:
```
@custom_jvp
def fn_key(key):
    return 3 + jax.random.normal(key) * 2

@fn_key.defjvp
def fn_key_jvp(primals, tangents):
    primal_out = fn_key(primals)
    tangent_out = 2
    return primal_out, tangent_out
```
and evaluate as follows:
```
key = jax.random.PRNGKey(42)
placeholder = np.zeros_like(key, dtype=dtypes.float0)
jax.jvp(fn_key, [key], [placeholder])
```
but I think I am doing something wrong, because it returns 0 as the gradient.

**Comment by sharadmv** [comment]
I don't think the gradient is well-defined for the `fn_key` because it only takes in a PRNGKey (which has an empty tangent type). What exactly do you want the gradient wrt in the `fn_key` case?

**Reply by mattjj** [reply]
We can probably identify this question with the usual "gradient with respect to intermediates" question (e.g. #5336, #6668). That is, there's nothing special about RVs here; it's just another intermediate.

Since JAX's autodiff is based around Python functions (identifying them with mathematical functions), you can only really differentiation with respect to a formal parameter of a Python function, and not an intermediate in its body. 

One trick you can use to generate a formal parameter with respect to which you can differentiate is the "perturbation function trick", where instead of:

```python
def fn_key(key):
    return 3 + jax.random.normal(key) * 2
```

you would write

```python
def fn_key(key, perturbation):
    return 3 + (jax.random.normal(key) + perturbation) * 2
```

and then differentiate `grad(fn_key, argnums=1)(key, 0.0)`.

See the thread of #5336 for more.

**Reply by pecey** [reply]
@sharadmv We wanted to take the grad wrt eps rather than key but we are working with a black box model if you will and the sampling is implicit in the function. As you pointed out, I understand that taking grad wrt key doesn't really make sense, and so I was just trying to see if it could be worked around.
Ah. That's a simple and clever trick. Thanks a lot @mattjj.

Also, maybe an extremely silly question but why did the `custom_jvp` definition not work and return 0? Any material that you could point to for understanding the internals of `custom_jvp`? 
Thank you.

**Reply by mattjj** [reply]
It's not silly! It's subtle.

The custom JVP rule isn't being called (try putting a print inside it), because
1. the tangent types for integer-dtype values (like `key`) are trivial in the sense that they represent the trivial vector space, which only has a zero vector in it, and
2. JAX assumes JVP rules are linear in their tangent input so that optimizations can be applied; in particular we assume that given a statically-known-to-be-zero input we'd get a zero output (so we don't need to call the rule at all).

Putting these together explains why the JVP rule isn't being called. Also it explains why this works:

```python
import numpy as np

import jax
from jax import custom_jvp
from jax import dtypes

@custom_jvp
def fn_key(key, param):
    del param  # doens't even need to be used!
    return 3 + jax.random.normal(key) * 2

@fn_key.defjvp
def fn_key_jvp(primals, tangents):
    primal_out = fn_key(*primals)
    tangent_out = 2.0
    return primal_out, tangent_out

key = jax.random.PRNGKey(42)
placeholder = np.zeros_like(key, dtype=dtypes.float0)
print(jax.jvp(fn_key, [key, 0.0], [placeholder, 0.0]))  # prints (..., 2.0)
```
